commit aa065263ecb0b0df096ff4b02b2fb88060dd2d45
Author: Gábor Stefanik <netrolller.3d@gmail.com>
Date:   Fri Aug 21 20:44:09 2009 +0200

    iwlwifi: Make injection of non-broadcast frames work again
    
    Commit 1ccb84d87d04df3c76cd4352fe69786d8c7cf016 by Wey-Yi Guy
    ("iwlwifi: clean up unused NL80211_IFTYPE_MONITOR for Monitor mode")
    broke injection of non-broadcast frames to unassociated stations
    (causing a SYSASSERT for all such injected frames), due to injected
    frames no longer automatically getting a broadcast station ID assigned.
    This patch restores the old behavior, fixing the aforementioned
    regression.
    
    Also, consistently check for IEEE80211_TX_CTL_INJECTED instead of
    iwl_is_monitor_mode in the TX path, as TX_CTL_INJECTED specifically
    means that a given packet is coming from a monitor interface, while
    iwl_is_monitor_mode only shows whether a monitor interface exists
    on the device.
    
    Signed-off-by: Gábor Stefanik <netrolller.3d@gmail.com>
    Acked-by: Reinette Chatre <reinette.chatre@intel.com>
    Signed-off-by: John W. Linville <linville@tuxdriver.com>

diff --git a/drivers/net/wireless/iwlwifi/iwl3945-base.c b/drivers/net/wireless/iwlwifi/iwl3945-base.c
index f339c5bd1fde..34790413eadd 100644
--- a/drivers/net/wireless/iwlwifi/iwl3945-base.c
+++ b/drivers/net/wireless/iwlwifi/iwl3945-base.c
@@ -518,9 +518,9 @@ static int iwl3945_tx_skb(struct iwl_priv *priv, struct sk_buff *skb)
 		IWL_DEBUG_TX(priv, "Sending REASSOC frame\n");
 #endif
 
-	/* drop all data frame if we are not associated */
+	/* drop all non-injected data frame if we are not associated */
 	if (ieee80211_is_data(fc) &&
-	    (!iwl_is_monitor_mode(priv)) && /* packet injection */
+	    !(info->flags & IEEE80211_TX_CTL_INJECTED) &&
 	    (!iwl_is_associated(priv) ||
 	     ((priv->iw_mode == NL80211_IFTYPE_STATION) && !priv->assoc_id))) {
 		IWL_DEBUG_DROP(priv, "Dropping - !iwl_is_associated\n");
@@ -532,7 +532,10 @@ static int iwl3945_tx_skb(struct iwl_priv *priv, struct sk_buff *skb)
 	hdr_len = ieee80211_hdrlen(fc);
 
 	/* Find (or create) index into station table for destination station */
-	sta_id = iwl_get_sta_id(priv, hdr);
+	if (info->flags & IEEE80211_TX_CTL_INJECTED)
+		sta_id = priv->hw_params.bcast_sta_id;
+	else
+		sta_id = iwl_get_sta_id(priv, hdr);
 	if (sta_id == IWL_INVALID_STATION) {
 		IWL_DEBUG_DROP(priv, "Dropping - INVALID STATION: %pM\n",
 			       hdr->addr1);